### ISSUE TITLE
`NetworkWeapon.fire()` is not within Inputs.

### ISSUE THREAD

**Issue by TheYellowArchitect** [foxssake/netfox#253]
```gdscript

#The below function is located Inside brawler scene, on the Weapon node
func _tick(_delta: float, _t: int):
	if input.is_firing:
		fire()
```

The function `fire()` directly changes state as it spawns a bomb projectile with velocity etc, yet the above function is exclusively a local input - it isn't shared to other players via RPCs, nor is it rollbacked.
This is a design flaw.

Especially noticeable if a replay manager is to be implemented, because this forces to capture all states, instead of all inputs and the very first state.

There are 2 ways to solve this problem:
1. `fire()` changes the state (same as current implementation), but also becomes a dictionary key to `_inputs`
I would suggest this key becoming `button_pressed` and its value to be an integer, so there can be mapped any amount of input buttons, for the same dictionary key. E.g. if there are 2 weapons to use, or more, this handles them all finely, with the condition that only 1 button can be pressed each tick (so you cannot fire both weapons at the same tick, the next weapon fire is queued for the next tick)

2. `fire()` changes the input directly, and the input itself is checked to spawn the bomb projectile. No idea if this will create bugs on existing code (by having the projectile spawn by state AND input, idk if there are proper checks already)
^This is already half-implemented, as the input is changed, it just isn't passed into `_inputs`

**Comment by TheYellowArchitect** [comment]
ok so I found out netfox also has input prediction/extrapolation. This function is in the class `rollback-synchronizer.gd`, in the following function:
```gdscript
func _get_history(buffer: Dictionary, tick: int) -> Dictionary:
	if buffer.has(tick):
		return buffer[tick]

	if buffer.is_empty():
		return {}
	
	var earliest = buffer.keys().min()
	var latest = buffer.keys().max()

	
	if tick < earliest:
		return buffer[earliest]
	
	#This is input prediction/extrapolation
	if tick > latest:
		return buffer[latest]
	
	var before = buffer.keys() \
		.filter(func (key): return key < tick) \
		.max()
	
	return buffer[before]
```

Currently, if `input.fire()` were placed in the input dictionary (`is_firing` is an input, but it's not in a dictionary or sent via an RPC), it wouldn't cause practical problems because cooldown wouldn't let it fire for the next ticks. But in a machinegun/flamethrower scenario, it would bug the game.

Kinda the same with jump not causing problems, if you are already in the air, it doesnt matter, and if you just pressed it a few ticks ago, you are already in the air, so the problem wasn't noticed thus far.

Basically, the input prediction/extrapolation should continue working as is, but every property should have a boolean "is extrapolateable" and if not, then have a default value (e.g. for movement, it would be Vector3.ZERO)

**Comment by elementbound** [comment]
> This is a design flaw.
How?

You want the server to know you want to fire, so you call an RPC. Since it's using an RPC, there's no point to also send a bool, since the RPC does all the work. It is also not part of the rollback process indeed - I kept the projectiles out of the rollback loop to keep the code simple. 

Could you please elaborate on why the above design should be changed? And also how your suggestions would solve the above? They seem to be concerned with *transmitting* the input, but not with the whole firing sequence ( client requests, server approves and broadcasts, client reconciles based on server response ).

---

> Turns out netfox already has input prediction/extrapolation!

Kind of. If you try to grab a state or input for a tick that we don't have data for, it will try to find *something* that being the latest data up to that point. However, during rollback, we don't resimulate ticks that we don't have input for.

**Comment by TheYellowArchitect** [comment]
>You want the server to know you want to fire, so you call an RPC. Since it's using an RPC, there's no point to also send a bool, since the RPC does all the work. It is also not part of the rollback process indeed - I kept the projectiles out of the rollback loop to keep the code simple.

The state changes without an input change. For example, a replay system cannot work by simply having the initial states, and all inputs. The input fire works locally, but isn't transmitted to others as a fire input.

>And also how your suggestions would solve the above? They seem to be concerned with transmitting the input, but not with the whole firing sequence ( client requests, server approves and broadcasts, client reconciles based on server response ).

_Using the 1st suggestion of the OP_:
Basically it becomes fully input, and the projectile is spawned for each player via input, but idk if this will break the logic of projectile reconcillation/rejection/acceptance.

_Using the 2nd suggestion of the OP_:
1. Player presses click
2. `_inputs[tick]` boolean becomes true (currently it was just `input.fire()`) and it changes the state by creating the projectile like currently, and also transmits this input. There is a flag or sth so other clients who don't own that boolean input don't try to create a state, and the game works as is.

_3rd suggestion not included in the OP_:
Everything is kept as is, but for the replay manager, i also store the creation of every projectile with its initial values. Kinda messes up the whole states/inputs separation, but its possible, though kinda hacky.

----------------

Minor argument I forgot: For a game with many weapons, it would be clean to contain their firing/button press in the inputs. In the current game I'm making, most actions are integer mapped, so instead of 1 boolean for each button, its basically 1 integer which contains exclusively 1 action. So if you press crouch and fire1 and fire2, it only includes one of these actions (crouch) and queues the next action (fire1 and fire2) for next tick(s) but perhaps this design isn't very compatible with netfox.

**Comment by TheYellowArchitect** [comment]
Ah also, another argument to include it as input is that by implementing [InputDelay](https://github.com/foxssake/netfox/pull/276) or InputSlack, the firing of a (bomb) projectile isn't included at all.

**Comment by elementbound** [comment]
I'll return to this issue with more detail, but I have to highlight this: 

> For example, a replay system cannot work by simply having the initial states, and all inputs.

**Currently there's no replay system. I won't optimize for systems that do not exist. Stop referring to systems that don't exist in netfox.**

**Comment by elementbound** [comment]
Since it would take some time until I can dedicate more attention to this, I suggest building whatever makes sense for your game.

What I'd probably do is this:
1. Tie firing to input as suggested
1. Spawn the projectiles in rollback tick
1. Cache them, so even if we rollback to this tick again, don't spawn more projectiles
1. Hook the projectiles into NetworkRollback events so they update their state

The last point would ideally be supported by a fully predictive version of RollbackSynchronizer, i.e. no state is synchronized, only recorded, trusting that each client has the same or similar enough simulation result.

The caching also lets you implement cancellation. So if during a rollback you find out that the player in fact couldn't have fired but you've already spawned the projectiles, you can just remove them.

**Comment by elementbound** [comment]
Closing in favor of #312 - both issues request rollback-managed projectiles. Can't give timelines, but I do plan on providing an example for this.
